test/client_test.exs

```diff
@@ -39,18 +39,74 @@ defmodule Sentry.ClientTest do
 
     Application.put_env(:sentry, :dsn, "http://public:secret@localhost:#{bypass.port}/1")
 
-      try do
-        Event.not_a_function
-      rescue
-        e ->
+    try do
+      Event.not_a_function
+    rescue
+      e ->
         assert capture_log(fn ->
           Sentry.capture_exception(e)
         end) =~ ~r/400.*Creation of this event was denied due to rate limiting/
-      end
+    end
   end
 
   test "errors when attempting to report invalid JSON" do
     unencodable_tuple = {:a, :b, :c}
     assert :error = Sentry.capture_message(unencodable_tuple)
   end
+
+  test "calls anonymous pre_event_send_function" do
+    bypass = Bypass.open
+    Bypass.expect bypass, fn conn ->
+      {:ok, body, conn} = Plug.Conn.read_body(conn)
+      request_map = Poison.decode!(body)
+      assert request_map["extra"] == %{"key" => "value"}
+      assert request_map["user"]["id"] == 1
+      Plug.Conn.resp(conn, 200, ~s<{"id": "340"}>)
+    end
+
+    Application.put_env(:sentry, :dsn, "http://public:secret@localhost:#{bypass.port}/1")
+    Logger.metadata([key: "value", user_id: 1])
+
+    Application.put_env(:sentry, :pre_event_send_function, fn(e) ->
+      metadata = Enum.into(Logger.metadata, %{})
+      {user_id, rest_metadata} = Map.pop(metadata, :user_id)
+      %{e | extra: Map.merge(e.extra, rest_metadata), user: Map.put(e.user, :id, user_id)}
+    end)
+
+    try do
+      Event.not_a_function
+    rescue
+      e ->
+        assert capture_log(fn ->
+          Sentry.capture_exception(e)
+        end)
+    end
+
+    Application.delete_env(:sentry, :pre_event_send_function)
```

Comment:
This line won't get reached if any thing above fails, which means that if this test fails, it "leaks" its mutations to the app env.  It's better to use an `on_exit` to define cleanup that will be guaranteed to run.  On my projects, we've tended to wrap this up in a helper function:

``` elixir
    def modify_env(app, overrides) do
      original_env = Application.get_all_env(app)
      Enum.each(overrides, fn {key, value} -> Application.put_env(app, key, value) end)

      ExUnit.Callbacks.on_exit(fn ->
        Enum.each overrides, fn {key, _} ->
          if Keyword.has_key?(original_env, key) do
            Application.put_env(app, key, Keyword.fetch!(original_env, key))
          else
            Application.delete_env(app, key)
          end
        end
      end)
    end
```

Then any test can use `modify_env` and trust it will get cleaned up correctly.